overlord/snapstate/helpers.go

```diff
@@ -0,0 +1,132 @@
+ *
+ */
+
+package snapstate
+
+import (
+	"errors"
+
+	"github.com/snapcore/snapd/overlord/restart"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+// Observed link order between essential snaps in doUpdate (snapstate.go)
+// base waits for snapd
+// gadget waits for base/os + snapd
+// kernel waits for base/os + snapd + gadget if present
+var essentialSnapsRestartOrder = []snap.Type{
+	snap.TypeOS,
+	// The base will only require restart if it's the base of the model.
+	snap.TypeBase,
+	snap.TypeGadget,
+	// Kernel must wait for gadget because the gadget may define
+	// new "$kernel:refs".
+	snap.TypeKernel,
+}
+
+func maybeSnapSetup(ts *state.TaskSet) *SnapSetup {
+	for _, t := range ts.Tasks() {
+		snapsup, err := TaskSnapSetup(t)
+		if err == nil {
+			return snapsup
+		}
+	}
+	return nil
+}
+
+func mappedTaskSets(tss []*state.TaskSet, bootBase string) (map[snap.Type]*state.TaskSet, error) {
+	avail := make(map[snap.Type]*state.TaskSet)
+	for _, ts := range tss {
+		snapsup := maybeSnapSetup(ts)
+		if snapsup == nil {
+			continue
+		}
+
+		switch snapsup.Type {
+		case snap.TypeBase:
+			if snapsup.SnapName() == bootBase {
+				avail[snapsup.Type] = ts
+			}
+		case snap.TypeOS, snap.TypeGadget, snap.TypeKernel:
+			avail[snapsup.Type] = ts
+		}
+	}
+	return avail, nil
+}
+
+func findUnlinkTask(ts *state.TaskSet) *state.Task {
+	for _, t := range ts.Tasks() {
+		switch t.Kind() {
+		case "unlink-snap", "unlink-current-snap":
+			return t
+		}
+	}
+	return nil
+}
+
+// SetDefaultRestartBoundaries marks "link-snap" (Do) and "unlink-snap"/"unlink-current-snap" (Undo) as restart boundaries
+// to maintain the old restart logic. This means that a restart must be performed after each of those tasks
+// in order for the change to continue.
+func SetDefaultRestartBoundaries(ts *state.TaskSet) {
+	linkSnap := ts.MaybeEdge(MaybeRebootEdge)
+	if linkSnap != nil {
+		restart.MarkTaskAsRestartBoundary(linkSnap, restart.RestartBoundaryDirectionDo)
+	}
+
+	unlinkSnap := findUnlinkTask(ts)
+	if unlinkSnap != nil {
+		restart.MarkTaskAsRestartBoundary(unlinkSnap, restart.RestartBoundaryDirectionUndo)
+	}
+}
+
+// SetEssentialSnapRestartBoundaries sets up default restart boundaries for a list of task-sets. If the
+// list of task-sets contain any updates/installs of essential snaps (base,gadget,kernel), then proper
+// restart boundaries will be set up for them.
+func SetEssentialSnapRestartBoundaries(tss []*state.TaskSet, bootBase string) error {
+	mappedTss, err := mappedTaskSets(tss, bootBase)
+	if err != nil {
+		return err
+	}
+
+	for _, o := range essentialSnapsRestartOrder {
+		if mappedTss[o] == nil {
+			continue
+		}
+		SetDefaultRestartBoundaries(mappedTss[o])
+	}
+	return nil
+}
+
+// DeviceModelBase returns the base-snap name of the current model. For UC16
+// this will return "core".
+func DeviceModelBase(st *state.State) (string, error) {
```

Comment:
it's unclear to me why this is exposed instead of the helpers needing this taking state, also BootBaseInfo already exists doing so this would need at least to be named consistently with that

Reply:
Good point about both. It definitely makes sense to let the other helpers take state instead!

Reply:
I ended up not doing this because of how it's used with remodelling, where we want to get the boot-base of the new device model, and not the current stored in state. I'll take any suggestions for improvement here. But instead of exposing it, it's now just a helper function as its only used inside the snapstate package.

Reply:
Thank you! I've updated the code